src/glossary.md

```diff
@@ -1,5 +1,29 @@
 # Glossary
 
+## Terms used in the Specification
+
+### Ill-formed Program
+
+A Rust Program is ill-formed if it violates a statically set constraint placed on it by the language.
+Unless otherwise statement explicitly an implementation shall issue an appropriate diagnostic for an ill-formed program and shall not translate such a program.
+
+#### No Diagnostic Required
+
+When No Diagnostic is Required for a static constraint, the implementation may issue a diagnostic for the violation, and may translate the program.
+If such a program is succesfully translated, no constraints are placed on the result of executing the program.
+
+> [!NOTE]
+> The constraint which is modified by the phrase "No Diagnostic Required" may be defined implicitly, such as with the term "shall", rather than explicitly stating that the program is ill-formed.
+
+> [!NOTE]
+> Such constraints are used only when it may be unreasonable or impossible to check the constraints statically, and where the result of violating the constraints makes it impossible to ensure the required behaviour of the program
+
+### Shall
+
+The word "shall" is a static constraint placed upon the program. A program that violates a constraint that uses "shall" is ill-formed.
```

Comment:
I don't entirely object to specifying this, but I just want to call out that it seems unusual to do so. At least, I have not seen any other languages do that. Is there a precedent that shows this needing definition or way this could be confusing otherwise?

Reply:
I thought C++ did but I checked and am incorrect.
I know C does so implicitly. I think it's useful to call out the term shall as specifically meaning a static constraint, because it could reasonably mean either.

Reply:
The Ada standards do (since Ada 95; Ada 83 used "must"). See eg [Ada 2012 §1.1.5:3][1].

[1]:http://ada-auth.org/standards/12rm/html/RM-1-1-5.html